src/tests/test_qasm.py

```diff
@@ -851,6 +852,21 @@ def test_execute_different_qat_input_types(self):
         with pytest.raises(TypeError):
             execute_qasm(qat_input=builder.instructions, hardware=hw)
 
+    def test_serialized_references_persist(self):
+        qasm_string = get_qasm2("serialize_orphan.qasm")
+        hardware = get_default_echo_hardware(8)
+        config = CompilerConfig()
+
+        frontend = QASMFrontend()
+        builder, metrics = frontend.parse(qasm_string, hardware, config)
+
+        serialized_builder = builder.serialize()
+        builder = QuantumInstructionBuilder.deserialize(serialized_builder)
+
+        results = frontend.execute(builder, hardware, config)
```

Comment:
the builder has a model/hardware on it. I think both cases should be tested here, that you use the original hardward and you also use the rehydrated hardware from the builder.